{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This pipeline assumes ARTIC V3 amplicon sequencing using ONT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import binascii\n",
    "import gzip\n",
    "import json\n",
    "import os\n",
    "import sys\n",
    "\n",
    "from io import BytesIO\n",
    "\n",
    "import pandas as pd\n",
    "import altair as alt\n",
    "\n",
    "from altair_saver import save\n",
    "#from Bio import SeqIO\n",
    "from IPython.display import HTML\n",
    "from onecodex import Api\n",
    "from onecodex.notebooks.report import set_style, title"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "ocx = Api()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "ENVIRONMENT = os.environ.get(\"ONE_CODEX_REPORT_ENV\", \"draft\")\n",
    "\n",
    "if ENVIRONMENT == \"production\":\n",
    "    sample_uuid = os.environ[\"ONE_CODEX_SAMPLE_UUID\"]\n",
    "    sample = ocx.Samples.get(sample_uuid)\n",
    "    assert sample is not None, \"Sample does not exist\"\n",
    "    sample_filename = sample.filename\n",
    "    if not os.path.exists(sample_filename):\n",
    "        sample.download()\n",
    "else:\n",
    "    sample_filename = \"sample.fastq\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# note our DB actually has v1 of this assembly as of Feb 2020\n",
    "VARIANTS_TSV_PATH = \"variants.tsv\"\n",
    "NEXTCLADE_TSV_PATH = \"nextclade.tsv\"\n",
    "PANGOLIN_CSV_PATH = \"pangolin.csv\"\n",
    "BAM_PATH = \"covid19.bam\"\n",
    "REFERENCE_PATH = os.environ.get(\n",
    "    \"FASTA_REFERENCE\", \"share/nCoV-2019.reference.fasta\"\n",
    ")\n",
    "#BED_FILE_PATH = os.environ.get(\"BED_FILE_PATH\", \"reference/artic-v3/ARTIC-V3.bed\")\n",
    "REFERENCE_NAME = os.path.basename(REFERENCE_PATH).rstrip('.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/\n",
      "Processing 1 files in ./\n",
      "guppyplexed_.fastq\t1345877\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22m artic-tools validate_scheme /artic-ncov2019/primer_schemes/nCoV-2019/V3/nCoV-2019.scheme.bed\n",
      "[02:20:36] [artic-tools::validate_scheme] starting primer scheme validator\n",
      "[02:20:36] [artic-tools::validate_scheme] reading scheme\n",
      "[02:20:36] [artic-tools::validate_scheme] collecting scheme stats\n",
      "[02:20:36] [artic-tools::validate_scheme] \tprimer scheme file:\t/artic-ncov2019/primer_schemes/nCoV-2019/V3/nCoV-2019.scheme.bed\n",
      "[02:20:36] [artic-tools::validate_scheme] \treference sequence:\tMN908947.3\n",
      "[02:20:36] [artic-tools::validate_scheme] \tnumber of pools:\t2\n",
      "[02:20:36] [artic-tools::validate_scheme] \tnumber of primers:\t218 (includes 22 alts)\n",
      "[02:20:36] [artic-tools::validate_scheme] \tminimum primer size:\t22\n",
      "[02:20:36] [artic-tools::validate_scheme] \tmaximum primer size:\t57\n",
      "[02:20:36] [artic-tools::validate_scheme] \tnumber of amplicons:\t98\n",
      "[02:20:36] [artic-tools::validate_scheme] \tmean amplicon size:\t343\n",
      "[02:20:36] [artic-tools::validate_scheme] \tmaximum amplicon size:\t375\n",
      "[02:20:36] [artic-tools::validate_scheme] \tscheme ref. span:\t30-29866\n",
      "[02:20:36] [artic-tools::validate_scheme] \tscheme overlaps:\t12.850247%\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22mminimap2 -a -x map-ont -t 6 /artic-ncov2019/primer_schemes/nCoV-2019/V3/nCoV-2019.reference.fasta sample_guppyplexed.fastq | samtools view -bS -F 4 - | samtools sort -o sample.sorted.bam -\n",
      "[M::mm_idx_gen::0.004*1.33] collected minimizers\n",
      "[M::mm_idx_gen::0.014*1.22] sorted minimizers\n",
      "[M::main::0.014*1.22] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.014*1.21] mid_occ = 3\n",
      "[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.015*1.20] distinct minimizers: 5587 (99.93% are singletons); average occurrences: 1.004; average spacing: 5.332\n",
      "[M::worker_pipeline::70.728*2.86] mapped 1158835 sequences\n",
      "[M::worker_pipeline::73.919*2.74] mapped 187042 sequences\n",
      "[M::main] Version: 2.17-r941\n",
      "[M::main] CMD: minimap2 -a -x map-ont -t 6 /artic-ncov2019/primer_schemes/nCoV-2019/V3/nCoV-2019.reference.fasta sample_guppyplexed.fastq\n",
      "[M::main] Real time: 73.967 sec; CPU: 202.795 sec; Peak RSS: 1.601 GB\n",
      "[bam_sort_core] merging from 1 files and 1 in-memory blocks...\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22msamtools index sample.sorted.bam\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22malign_trim --normalise 200 /artic-ncov2019/primer_schemes/nCoV-2019/V3/nCoV-2019.scheme.bed --start --remove-incorrect-pairs --report sample.alignreport.txt < sample.sorted.bam 2> sample.alignreport.er | samtools sort -T sample - -o sample.trimmed.rg.sorted.bam\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22malign_trim --normalise 200 /artic-ncov2019/primer_schemes/nCoV-2019/V3/nCoV-2019.scheme.bed --remove-incorrect-pairs --report sample.alignreport.txt < sample.sorted.bam 2> sample.alignreport.er | samtools sort -T sample - -o sample.primertrimmed.rg.sorted.bam\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22msamtools index sample.trimmed.rg.sorted.bam\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22msamtools index sample.primertrimmed.rg.sorted.bam\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22mmedaka consensus --model r941_min_high_g360 --threads 6 --chunk_len 800 --chunk_ovlp 400 --RG nCoV-2019_2 sample.trimmed.rg.sorted.bam sample.nCoV-2019_2.hdf\n",
      "[02:26:50 - medaka] Reads will be filtered to only those with RG tag: nCoV-2019_2\n",
      "[02:26:50 - Predict] Processing region(s): MN908947.3:0-29903\n",
      "[02:26:50 - Predict] Setting tensorflow threads to 6.\n",
      "[02:26:50 - Predict] Processing 1 long region(s) with batching.\n",
      "[02:26:50 - Predict] Using model: /opt/conda/envs/artic-ncov2019/lib/python3.6/site-packages/medaka/data/r941_min_high_g360_model.hdf5.\n",
      "[02:26:50 - ModelLoad] Building model with cudnn optimization: False\n",
      "[02:26:51 - DLoader] Initializing data loader\n",
      "[02:26:51 - PWorker] Running inference for 0.0M draft bases.\n",
      "[02:26:51 - Sampler] Initializing sampler for consensus of region MN908947.3:0-29903.\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 320-725.\n",
      "[02:26:51 - Feature] Processed MN908947.3:320.0-725.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 943-1336.\n",
      "[02:26:51 - Feature] Processed MN908947.3:943.0-1336.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 1573-1963.\n",
      "[02:26:51 - Feature] Processed MN908947.3:1573.0-1963.2 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 2181-2591.\n",
      "[02:26:51 - Feature] Processed MN908947.3:2181.0-2591.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 2826-3209.\n",
      "[02:26:51 - Feature] Processed MN908947.3:2826.0-3209.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 3460-3852.\n",
      "[02:26:51 - Feature] Processed MN908947.3:3460.0-3852.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 4051-4449.\n",
      "[02:26:51 - Feature] Processed MN908947.3:4051.0-4449.0 (median depth 86.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 4636-5016.\n",
      "[02:26:51 - Feature] Processed MN908947.3:4636.0-5016.1 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 5867-6271.\n",
      "[02:26:51 - Feature] Processed MN908947.3:5867.0-6271.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 6466-6872.\n",
      "[02:26:51 - Feature] Processed MN908947.3:6466.0-6872.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 7035-7414.\n",
      "[02:26:51 - Feature] Processed MN908947.3:7035.0-7414.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 7626-8018.\n",
      "[02:26:51 - Feature] Processed MN908947.3:7626.0-8018.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 8249-8660.\n",
      "[02:26:51 - Feature] Processed MN908947.3:8249.0-8660.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 8888-9270.\n",
      "[02:26:51 - Feature] Processed MN908947.3:8888.0-9270.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 9477-9857.\n",
      "[02:26:51 - Feature] Processed MN908947.3:9477.0-9857.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 10076-10458.\n",
      "[02:26:51 - Feature] Processed MN908947.3:10076.0-10458.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 10666-11073.\n",
      "[02:26:51 - Feature] Processed MN908947.3:10666.0-11073.0 (median depth 390.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 11306-11692.\n",
      "[02:26:51 - Feature] Processed MN908947.3:11306.0-11692.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 11863-12255.\n",
      "[02:26:51 - Feature] Processed MN908947.3:11863.0-12255.1 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 12417-12801.\n",
      "[02:26:51 - Feature] Processed MN908947.3:12417.0-12801.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 13005-13399.\n",
      "[02:26:51 - Feature] Processed MN908947.3:13005.0-13399.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 13599-13983.\n",
      "[02:26:51 - Feature] Processed MN908947.3:13599.0-13983.1 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 14207-14600.\n",
      "[02:26:51 - Feature] Processed MN908947.3:14207.0-14600.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 14865-15245.\n",
      "[02:26:51 - Feature] Processed MN908947.3:14865.0-15245.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 15481-15885.\n",
      "[02:26:51 - Feature] Processed MN908947.3:15481.0-15885.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 16118-16509.\n",
      "[02:26:51 - Feature] Processed MN908947.3:16118.0-16509.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 16748-17151.\n",
      "[02:26:51 - Feature] Processed MN908947.3:16748.0-17151.1 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 17381-17760.\n",
      "[02:26:51 - Feature] Processed MN908947.3:17381.0-17760.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 17966-18347.\n",
      "[02:26:51 - Feature] Processed MN908947.3:17966.0-18347.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 18596-18978.\n",
      "[02:26:51 - Feature] Processed MN908947.3:18596.0-18978.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 19204-19615.\n",
      "[02:26:51 - Feature] Processed MN908947.3:19204.0-19615.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 19844-20254.\n",
      "[02:26:51 - Feature] Processed MN908947.3:19844.0-20254.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 20472-20889.\n",
      "[02:26:51 - Feature] Processed MN908947.3:20472.0-20889.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 21075-21454.\n",
      "[02:26:51 - Feature] Processed MN908947.3:21075.0-21454.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 21658-22037.\n",
      "[02:26:51 - Feature] Processed MN908947.3:21658.0-22037.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 22262-22649.\n",
      "[02:26:51 - Feature] Processed MN908947.3:22262.0-22649.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 23443-23846.\n",
      "[02:26:51 - Feature] Processed MN908947.3:23443.0-23846.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 24078-24466.\n",
      "[02:26:51 - Feature] Processed MN908947.3:24078.0-24466.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 24696-25075.\n",
      "[02:26:51 - Feature] Processed MN908947.3:24696.0-25075.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 25279-25672.\n",
      "[02:26:51 - Feature] Processed MN908947.3:25279.0-25672.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 25902-26314.\n",
      "[02:26:51 - Feature] Processed MN908947.3:25902.0-26314.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 26520-26912.\n",
      "[02:26:51 - Feature] Processed MN908947.3:26520.0-26912.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 27141-27532.\n",
      "[02:26:51 - Feature] Processed MN908947.3:27141.0-27532.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 27784-28171.\n",
      "[02:26:51 - Feature] Processed MN908947.3:27784.0-28171.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 28394-28778.\n",
      "[02:26:51 - Feature] Processed MN908947.3:28394.0-28778.0 (median depth 400.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 28985-29377.\n",
      "[02:26:51 - Feature] Processed MN908947.3:28985.0-29377.0 (median depth 15.0)\n",
      "[02:26:51 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 29486-29865.\n",
      "[02:26:51 - Feature] Processed MN908947.3:29486.0-29865.0 (median depth 101.0)\n",
      "[02:26:51 - Sampler] Took 0.27s to make features.\n",
      "[02:26:51 - Sampler] Region MN908947.3:28985.0-29377.0 (488 positions) is smaller than inference chunk length 800, quarantining.\n",
      "[02:26:54 - PWorker] All done, 1 remainder regions.\n",
      "[02:26:54 - Predict] Processing 1 short region(s).\n",
      "[02:26:54 - ModelLoad] Building model with cudnn optimization: False\n",
      "[02:26:55 - DLoader] Initializing data loader\n",
      "[02:26:55 - PWorker] Running inference for 0.0M draft bases.\n",
      "[02:26:55 - Sampler] Initializing sampler for consensus of region MN908947.3:28985-29378.\n",
      "[02:26:55 - Feature] Processed MN908947.3:28985.0-29377.0 (median depth 15.0)\n",
      "[02:26:55 - Sampler] Took 0.06s to make features.\n",
      "[02:26:55 - PWorker] All done, 0 remainder regions.\n",
      "[02:26:55 - Predict] Finished processing all regions.\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22mmedaka variant /artic-ncov2019/primer_schemes/nCoV-2019/V3/nCoV-2019.reference.fasta sample.nCoV-2019_2.hdf sample.nCoV-2019_2.vcf\n",
      "[02:26:57 - DataIndex] Loaded 1/1 (100.00%) sample files.\n",
      "[02:26:57 - Variants] Processing MN908947.3:0-.\n",
      "[02:26:57 - TrimOverlap] MN908947.3:486.2-725.0 and MN908947.3:943.0-1190.1 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:57 - TrimOverlap] MN908947.3:1093.2-1336.0 and MN908947.3:1573.0-1822.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:57 - TrimOverlap] MN908947.3:1717.2-1963.2 and MN908947.3:2181.0-2422.3 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:57 - TrimOverlap] MN908947.3:2336.1-2591.0 and MN908947.3:2826.0-3066.3 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:57 - TrimOverlap] MN908947.3:2952.0-3209.0 and MN908947.3:3460.0-3687.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:57 - TrimOverlap] MN908947.3:3612.3-3852.0 and MN908947.3:4051.0-4444.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:57 - TrimOverlap] MN908947.3:4056.0-4449.0 and MN908947.3:4636.0-4869.1 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:57 - TrimOverlap] MN908947.3:4782.3-5016.1 and MN908947.3:5867.0-6099.2 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:57 - TrimOverlap] MN908947.3:6033.1-6271.0 and MN908947.3:6466.0-6718.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:57 - TrimOverlap] MN908947.3:6634.2-6872.0 and MN908947.3:7035.0-7280.2 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:57 - TrimOverlap] MN908947.3:7167.1-7414.0 and MN908947.3:7626.0-7883.2 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:57 - TrimOverlap] MN908947.3:7781.2-8018.0 and MN908947.3:8249.0-8490.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:57 - TrimOverlap] MN908947.3:8428.0-8660.0 and MN908947.3:8888.0-9120.2 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:57 - TrimOverlap] MN908947.3:9045.4-9270.0 and MN908947.3:9477.0-9717.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:57 - TrimOverlap] MN908947.3:9611.2-9857.0 and MN908947.3:10076.0-10301.1 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:57 - TrimOverlap] MN908947.3:10235.1-10458.0 and MN908947.3:10666.0-10923.1 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:57 - TrimOverlap] MN908947.3:10822.2-11073.0 and MN908947.3:11306.0-11553.3 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:57 - TrimOverlap] MN908947.3:11439.4-11692.0 and MN908947.3:11863.0-12097.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:57 - TrimOverlap] MN908947.3:12025.3-12255.1 and MN908947.3:12417.0-12664.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:57 - TrimOverlap] MN908947.3:12563.0-12801.0 and MN908947.3:13005.0-13233.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:57 - TrimOverlap] MN908947.3:13163.7-13399.0 and MN908947.3:13599.0-13842.2 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:57 - TrimOverlap] MN908947.3:13749.4-13983.1 and MN908947.3:14207.0-14451.2 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:57 - TrimOverlap] MN908947.3:14350.2-14600.0 and MN908947.3:14865.0-15095.5 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:57 - TrimOverlap] MN908947.3:15016.3-15245.0 and MN908947.3:15481.0-15703.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:57 - TrimOverlap] MN908947.3:15643.5-15885.0 and MN908947.3:16118.0-16356.1 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:57 - TrimOverlap] MN908947.3:16261.1-16509.0 and MN908947.3:16748.0-16987.3 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:58 - TrimOverlap] MN908947.3:16903.1-17151.1 and MN908947.3:17381.0-17618.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:58 - TrimOverlap] MN908947.3:17520.1-17760.0 and MN908947.3:17966.0-18190.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:58 - TrimOverlap] MN908947.3:18111.1-18347.0 and MN908947.3:18596.0-18821.1 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:58 - TrimOverlap] MN908947.3:18761.0-18978.0 and MN908947.3:19204.0-19454.1 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:58 - TrimOverlap] MN908947.3:19369.4-19615.0 and MN908947.3:19844.0-20088.2 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:58 - TrimOverlap] MN908947.3:20018.1-20254.0 and MN908947.3:20472.0-20729.3 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:58 - TrimOverlap] MN908947.3:20649.2-20889.0 and MN908947.3:21075.0-21325.2 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:58 - TrimOverlap] MN908947.3:21209.2-21454.0 and MN908947.3:21658.0-21885.2 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:58 - TrimOverlap] MN908947.3:21811.4-22037.0 and MN908947.3:22262.0-22499.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:58 - TrimOverlap] MN908947.3:22389.1-22649.0 and MN908947.3:23443.0-23669.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:58 - TrimOverlap] MN908947.3:23609.1-23846.0 and MN908947.3:24078.0-24318.3 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:58 - TrimOverlap] MN908947.3:24234.1-24466.0 and MN908947.3:24696.0-24924.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:58 - TrimOverlap] MN908947.3:24847.0-25075.0 and MN908947.3:25279.0-25513.5 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:58 - TrimOverlap] MN908947.3:25434.2-25672.0 and MN908947.3:25902.0-26144.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:58 - TrimOverlap] MN908947.3:26069.3-26314.0 and MN908947.3:26520.0-26775.4 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:58 - TrimOverlap] MN908947.3:26677.0-26912.0 and MN908947.3:27141.0-27398.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:58 - TrimOverlap] MN908947.3:27282.0-27532.0 and MN908947.3:27784.0-28029.2 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:58 - TrimOverlap] MN908947.3:27914.2-28171.0 and MN908947.3:28394.0-28643.4 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:58 - TrimOverlap] MN908947.3:28529.0-28778.0 and MN908947.3:28985.0-29377.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:26:58 - TrimOverlap] MN908947.3:28985.0-29377.0 and MN908947.3:29486.0-29843.1 cannot be concatenated as there is no overlap and they do not abut.\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22mmedaka tools annotate --pad 25 --RG nCoV-2019_2 sample.nCoV-2019_2.vcf /artic-ncov2019/primer_schemes/nCoV-2019/V3/nCoV-2019.reference.fasta sample.trimmed.rg.sorted.bam tmp.medaka-annotate.vcf\n",
      "[02:27:00 - medaka] Reads will be filtered to only those with RG tag: nCoV-2019_2\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22mmv tmp.medaka-annotate.vcf sample.nCoV-2019_2.vcf\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22mmedaka consensus --model r941_min_high_g360 --threads 6 --chunk_len 800 --chunk_ovlp 400 --RG nCoV-2019_1 sample.trimmed.rg.sorted.bam sample.nCoV-2019_1.hdf\n",
      "[02:27:01 - medaka] Reads will be filtered to only those with RG tag: nCoV-2019_1\n",
      "[02:27:01 - Predict] Processing region(s): MN908947.3:0-29903\n",
      "[02:27:01 - Predict] Setting tensorflow threads to 6.\n",
      "[02:27:01 - Predict] Processing 1 long region(s) with batching.\n",
      "[02:27:01 - Predict] Using model: /opt/conda/envs/artic-ncov2019/lib/python3.6/site-packages/medaka/data/r941_min_high_g360_model.hdf5.\n",
      "[02:27:01 - ModelLoad] Building model with cudnn optimization: False\n",
      "[02:27:02 - DLoader] Initializing data loader\n",
      "[02:27:02 - Sampler] Initializing sampler for consensus of region MN908947.3:0-29903.\n",
      "[02:27:02 - PWorker] Running inference for 0.0M draft bases.\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 30-409.\n",
      "[02:27:03 - Feature] Processed MN908947.3:30.0-409.0 (median depth 16.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 642-1027.\n",
      "[02:27:03 - Feature] Processed MN908947.3:642.0-1027.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 1242-1650.\n",
      "[02:27:03 - Feature] Processed MN908947.3:1242.0-1650.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 1871-2268.\n",
      "[02:27:03 - Feature] Processed MN908947.3:1871.0-2268.0 (median depth 77.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 2504-2903.\n",
      "[02:27:03 - Feature] Processed MN908947.3:2504.0-2903.0 (median depth 47.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 3144-3530.\n",
      "[02:27:03 - Feature] Processed MN908947.3:3144.0-3530.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 3771-4163.\n",
      "[02:27:03 - Feature] Processed MN908947.3:3771.0-4163.1 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 4294-4695.\n",
      "[02:27:03 - Feature] Processed MN908947.3:4294.0-4695.0 (median depth 138.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 4939-5320.\n",
      "[02:27:03 - Feature] Processed MN908947.3:4939.0-5320.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 5563-5956.\n",
      "[02:27:03 - Feature] Processed MN908947.3:5563.0-5956.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 6167-6549.\n",
      "[02:27:03 - Feature] Processed MN908947.3:6167.0-6549.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 6718-7116.\n",
      "[02:27:03 - Feature] Processed MN908947.3:6718.0-7116.1 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 7305-7693.\n",
      "[02:27:03 - Feature] Processed MN908947.3:7305.0-7693.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 7943-8340.\n",
      "[02:27:03 - Feature] Processed MN908947.3:7943.0-8340.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 8595-8982.\n",
      "[02:27:03 - Feature] Processed MN908947.3:8595.0-8982.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 9204-9584.\n",
      "[02:27:03 - Feature] Processed MN908947.3:9204.0-9584.0 (median depth 58.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 9784-10170.\n",
      "[02:27:03 - Feature] Processed MN908947.3:9784.0-10170.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 10362-10762.\n",
      "[02:27:03 - Feature] Processed MN908947.3:10362.0-10762.1 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 10999-11393.\n",
      "[02:27:03 - Feature] Processed MN908947.3:10999.0-11393.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 11555-11948.\n",
      "[02:27:03 - Feature] Processed MN908947.3:11555.0-11948.2 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 12110-12489.\n",
      "[02:27:03 - Feature] Processed MN908947.3:12110.0-12489.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 12710-13095.\n",
      "[02:27:03 - Feature] Processed MN908947.3:12710.0-13095.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 13318-13698.\n",
      "[02:27:03 - Feature] Processed MN908947.3:13318.0-13698.2 (median depth 183.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 13918-14298.\n",
      "[02:27:03 - Feature] Processed MN908947.3:13918.0-14298.1 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 14545-14925.\n",
      "[02:27:03 - Feature] Processed MN908947.3:14545.0-14925.1 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 15171-15559.\n",
      "[02:27:03 - Feature] Processed MN908947.3:15171.0-15559.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 15827-16208.\n",
      "[02:27:03 - Feature] Processed MN908947.3:15827.0-16208.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 16416-16832.\n",
      "[02:27:03 - Feature] Processed MN908947.3:16416.0-16832.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 17065-17451.\n",
      "[02:27:03 - Feature] Processed MN908947.3:17065.0-17451.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 17674-18061.\n",
      "[02:27:03 - Feature] Processed MN908947.3:17674.0-18061.0 (median depth 88.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 18253-18671.\n",
      "[02:27:03 - Feature] Processed MN908947.3:18253.0-18671.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 18896-19296.\n",
      "[02:27:03 - Feature] Processed MN908947.3:18896.0-19296.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 19548-19938.\n",
      "[02:27:03 - Feature] Processed MN908947.3:19548.0-19938.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 20172-20571.\n",
      "[02:27:03 - Feature] Processed MN908947.3:20172.0-20571.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 20786-21168.\n",
      "[02:27:03 - Feature] Processed MN908947.3:20786.0-21168.2 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 21357-21742.\n",
      "[02:27:03 - Feature] Processed MN908947.3:21357.0-21742.2 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 21961-22345.\n",
      "[02:27:03 - Feature] Processed MN908947.3:21961.0-22345.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 22516-22902.\n",
      "[02:27:03 - Feature] Processed MN908947.3:22516.0-22902.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 23122-23521.\n",
      "[02:27:03 - Feature] Processed MN908947.3:23122.0-23521.1 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 23789-24168.\n",
      "[02:27:03 - Feature] Processed MN908947.3:23789.0-24168.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 24391-24788.\n",
      "[02:27:03 - Feature] Processed MN908947.3:24391.0-24788.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 24978-25368.\n",
      "[02:27:03 - Feature] Processed MN908947.3:24978.0-25368.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 25601-25993.\n",
      "[02:27:03 - Feature] Processed MN908947.3:25601.0-25993.2 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 26197-26589.\n",
      "[02:27:03 - Feature] Processed MN908947.3:26197.0-26589.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 26835-27226.\n",
      "[02:27:03 - Feature] Processed MN908947.3:26835.0-27226.0 (median depth 40.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 27446-27853.\n",
      "[02:27:03 - Feature] Processed MN908947.3:27446.0-27853.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 28081-28463.\n",
      "[02:27:03 - Feature] Processed MN908947.3:28081.0-28463.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 28677-29062.\n",
      "[02:27:03 - Feature] Processed MN908947.3:28677.0-29062.0 (median depth 400.0)\n",
      "[02:27:03 - Feature] Pileup counts do not span requested region, requested MN908947.3:0-29903, received 29288-29692.\n",
      "[02:27:03 - Feature] Processed MN908947.3:29288.0-29692.0 (median depth 400.0)\n",
      "[02:27:03 - Sampler] Took 0.27s to make features.\n",
      "[02:27:03 - Sampler] Region MN908947.3:30.0-409.0 (486 positions) is smaller than inference chunk length 800, quarantining.\n",
      "[02:27:03 - Sampler] Region MN908947.3:1871.0-2268.0 (728 positions) is smaller than inference chunk length 800, quarantining.\n",
      "[02:27:03 - Sampler] Region MN908947.3:2504.0-2903.0 (700 positions) is smaller than inference chunk length 800, quarantining.\n",
      "[02:27:03 - Sampler] Region MN908947.3:9204.0-9584.0 (725 positions) is smaller than inference chunk length 800, quarantining.\n",
      "[02:27:03 - Sampler] Region MN908947.3:17674.0-18061.0 (799 positions) is smaller than inference chunk length 800, quarantining.\n",
      "[02:27:03 - Sampler] Region MN908947.3:26835.0-27226.0 (660 positions) is smaller than inference chunk length 800, quarantining.\n",
      "[02:27:06 - PWorker] All done, 6 remainder regions.\n",
      "[02:27:06 - Predict] Processing 6 short region(s).\n",
      "[02:27:06 - ModelLoad] Building model with cudnn optimization: False\n",
      "[02:27:06 - DLoader] Initializing data loader\n",
      "[02:27:06 - Sampler] Initializing sampler for consensus of region MN908947.3:30-410.\n",
      "[02:27:06 - PWorker] Running inference for 0.0M draft bases.\n",
      "[02:27:06 - Feature] Processed MN908947.3:30.0-409.0 (median depth 16.0)\n",
      "[02:27:06 - Sampler] Took 0.00s to make features.\n",
      "[02:27:07 - PWorker] All done, 0 remainder regions.\n",
      "[02:27:07 - DLoader] Initializing data loader\n",
      "[02:27:07 - PWorker] Running inference for 0.0M draft bases.\n",
      "[02:27:07 - Sampler] Initializing sampler for consensus of region MN908947.3:1871-2269.\n",
      "[02:27:07 - Feature] Processed MN908947.3:1871.0-2268.0 (median depth 77.0)\n",
      "[02:27:07 - Sampler] Took 0.01s to make features.\n",
      "[02:27:07 - PWorker] All done, 0 remainder regions.\n",
      "[02:27:07 - DLoader] Initializing data loader\n",
      "[02:27:07 - PWorker] Running inference for 0.0M draft bases.\n",
      "[02:27:07 - Sampler] Initializing sampler for consensus of region MN908947.3:2504-2904.\n",
      "[02:27:07 - Feature] Processed MN908947.3:2504.0-2903.0 (median depth 47.0)\n",
      "[02:27:07 - Sampler] Took 0.01s to make features.\n",
      "[02:27:07 - PWorker] All done, 0 remainder regions.\n",
      "[02:27:07 - DLoader] Initializing data loader\n",
      "[02:27:07 - PWorker] Running inference for 0.0M draft bases.\n",
      "[02:27:07 - Sampler] Initializing sampler for consensus of region MN908947.3:9204-9585.\n",
      "[02:27:07 - Feature] Processed MN908947.3:9204.0-9584.0 (median depth 58.0)\n",
      "[02:27:07 - Sampler] Took 0.03s to make features.\n",
      "[02:27:08 - PWorker] All done, 0 remainder regions.\n",
      "[02:27:08 - DLoader] Initializing data loader\n",
      "[02:27:08 - PWorker] Running inference for 0.0M draft bases.\n",
      "[02:27:08 - Sampler] Initializing sampler for consensus of region MN908947.3:17674-18062.\n",
      "[02:27:08 - Feature] Processed MN908947.3:17674.0-18061.0 (median depth 88.0)\n",
      "[02:27:08 - Sampler] Took 0.01s to make features.\n",
      "[02:27:08 - PWorker] All done, 0 remainder regions.\n",
      "[02:27:08 - DLoader] Initializing data loader\n",
      "[02:27:08 - PWorker] Running inference for 0.0M draft bases.\n",
      "[02:27:08 - Sampler] Initializing sampler for consensus of region MN908947.3:26835-27227.\n",
      "[02:27:08 - Feature] Processed MN908947.3:26835.0-27226.0 (median depth 40.0)\n",
      "[02:27:08 - Sampler] Took 0.04s to make features.\n",
      "[02:27:08 - PWorker] All done, 0 remainder regions.\n",
      "[02:27:08 - Predict] Finished processing all regions.\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22mmedaka variant /artic-ncov2019/primer_schemes/nCoV-2019/V3/nCoV-2019.reference.fasta sample.nCoV-2019_1.hdf sample.nCoV-2019_1.vcf\n",
      "[02:27:10 - DataIndex] Loaded 1/1 (100.00%) sample files.\n",
      "[02:27:10 - Variants] Processing MN908947.3:0-.\n",
      "[02:27:10 - TrimOverlap] MN908947.3:30.0-409.0 and MN908947.3:642.0-877.2 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:10 - TrimOverlap] MN908947.3:802.1-1027.0 and MN908947.3:1242.0-1469.2 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:10 - TrimOverlap] MN908947.3:1416.1-1650.0 and MN908947.3:1871.0-2268.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:10 - TrimOverlap] MN908947.3:1871.0-2268.0 and MN908947.3:2504.0-2903.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:10 - TrimOverlap] MN908947.3:2504.0-2903.0 and MN908947.3:3144.0-3394.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:10 - TrimOverlap] MN908947.3:3280.2-3530.0 and MN908947.3:3771.0-4040.2 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:10 - TrimOverlap] MN908947.3:3892.1-4163.1 and MN908947.3:4294.0-4617.1 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:10 - TrimOverlap] MN908947.3:4382.0-4695.0 and MN908947.3:4939.0-5170.1 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:10 - TrimOverlap] MN908947.3:5078.2-5320.0 and MN908947.3:5563.0-5777.1 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:10 - TrimOverlap] MN908947.3:5712.2-5956.0 and MN908947.3:6167.0-6413.1 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:10 - TrimOverlap] MN908947.3:6300.0-6549.0 and MN908947.3:6718.0-6948.2 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:10 - TrimOverlap] MN908947.3:6870.1-7116.1 and MN908947.3:7305.0-7555.1 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:10 - TrimOverlap] MN908947.3:7442.1-7693.0 and MN908947.3:7943.0-8167.3 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:10 - TrimOverlap] MN908947.3:8102.1-8340.0 and MN908947.3:8595.0-8832.2 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:10 - TrimOverlap] MN908947.3:8734.1-8982.0 and MN908947.3:9204.0-9584.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:10 - TrimOverlap] MN908947.3:9204.0-9584.0 and MN908947.3:9784.0-10000.3 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:10 - TrimOverlap] MN908947.3:9936.0-10170.0 and MN908947.3:10362.0-10606.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:10 - TrimOverlap] MN908947.3:10519.3-10762.1 and MN908947.3:10999.0-11234.3 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:10 - TrimOverlap] MN908947.3:11169.1-11393.0 and MN908947.3:11555.0-11787.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:10 - TrimOverlap] MN908947.3:11719.0-11948.2 and MN908947.3:12110.0-12349.1 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:10 - TrimOverlap] MN908947.3:12248.3-12489.0 and MN908947.3:12710.0-12955.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:12852.0-13095.0 and MN908947.3:13318.0-13608.1 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:13392.0-13698.2 and MN908947.3:13918.0-14158.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:14066.2-14298.1 and MN908947.3:14545.0-14795.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:14675.2-14925.1 and MN908947.3:15171.0-15422.1 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:15320.2-15559.0 and MN908947.3:15827.0-16079.2 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:15956.1-16208.0 and MN908947.3:16416.0-16655.1 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:16596.0-16832.0 and MN908947.3:17065.0-17292.2 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:17234.0-17451.0 and MN908947.3:17674.0-18061.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:17674.0-18061.0 and MN908947.3:18253.0-18482.1 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:18443.0-18671.0 and MN908947.3:18896.0-19132.4 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:19056.0-19296.0 and MN908947.3:19548.0-19793.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:19699.1-19938.0 and MN908947.3:20172.0-20409.1 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:20324.4-20571.0 and MN908947.3:20786.0-21023.3 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:20916.1-21168.2 and MN908947.3:21357.0-21609.1 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:21489.1-21742.2 and MN908947.3:21961.0-22194.1 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:22145.0-22345.0 and MN908947.3:22516.0-22748.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:22657.1-22902.0 and MN908947.3:23122.0-23361.24 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:23291.2-23521.1 and MN908947.3:23789.0-24016.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:23941.1-24168.0 and MN908947.3:24391.0-24634.1 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:24542.2-24788.0 and MN908947.3:24978.0-25229.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:25121.3-25368.0 and MN908947.3:25601.0-25847.2 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:25744.1-25993.2 and MN908947.3:26197.0-26441.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:26331.2-26589.0 and MN908947.3:26835.0-27226.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:26835.0-27226.0 and MN908947.3:27446.0-27695.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:27602.2-27853.0 and MN908947.3:28081.0-28326.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:28226.0-28463.0 and MN908947.3:28677.0-28935.1 cannot be concatenated as there is no overlap and they do not abut.\n",
      "[02:27:11 - TrimOverlap] MN908947.3:28804.2-29062.0 and MN908947.3:29288.0-29539.0 cannot be concatenated as there is no overlap and they do not abut.\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22mmedaka tools annotate --pad 25 --RG nCoV-2019_1 sample.nCoV-2019_1.vcf /artic-ncov2019/primer_schemes/nCoV-2019/V3/nCoV-2019.reference.fasta sample.trimmed.rg.sorted.bam tmp.medaka-annotate.vcf\n",
      "[02:27:13 - medaka] Reads will be filtered to only those with RG tag: nCoV-2019_1\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22mmv tmp.medaka-annotate.vcf sample.nCoV-2019_1.vcf\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22martic_vcf_merge sample /artic-ncov2019/primer_schemes/nCoV-2019/V3/nCoV-2019.scheme.bed 2> sample.primersitereport.txt nCoV-2019_2:sample.nCoV-2019_2.vcf nCoV-2019_1:sample.nCoV-2019_1.vcf\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22mbgzip -f sample.merged.vcf\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22mtabix -p vcf sample.merged.vcf.gz\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22martic-tools check_vcf --dropPrimerVars --dropOverlapFails --vcfOut sample.merged.filtered.vcf sample.merged.vcf.gz /artic-ncov2019/primer_schemes/nCoV-2019/V3/nCoV-2019.scheme.bed 2> sample.vcfreport.txt\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22mmv sample.merged.filtered.vcf sample.merged.vcf\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22martic_vcf_filter --medaka sample.merged.vcf sample.pass.vcf sample.fail.vcf\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22mbgzip -f sample.pass.vcf\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22mtabix -p vcf sample.pass.vcf.gz\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22martic_make_depth_mask --store-rg-depths /artic-ncov2019/primer_schemes/nCoV-2019/V3/nCoV-2019.reference.fasta sample.primertrimmed.rg.sorted.bam sample.coverage_mask.txt\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22martic_mask /artic-ncov2019/primer_schemes/nCoV-2019/V3/nCoV-2019.reference.fasta sample.coverage_mask.txt sample.fail.vcf sample.preconsensus.fasta\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22mbcftools consensus -f sample.preconsensus.fasta sample.pass.vcf.gz -m sample.coverage_mask.txt -o sample.consensus.fasta\n",
      "Note: the --sample option not given, applying all records regardless of the genotype\n",
      "Applied 2 variants\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22martic_fasta_header sample.consensus.fasta \"sample/ARTIC/medaka\"\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22mcat sample.consensus.fasta /artic-ncov2019/primer_schemes/nCoV-2019/V3/nCoV-2019.reference.fasta > sample.muscle.in.fasta\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22mmuscle -in sample.muscle.in.fasta -out sample.muscle.out.fasta\n",
      "\n",
      "MUSCLE v3.8.1551 by Robert C. Edgar\n",
      "\n",
      "http://www.drive5.com/muscle\n",
      "This software is donated to the public domain.\n",
      "Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\n",
      "\n",
      "sample.muscle.in 2 seqs, lengths min 29903, max 29903, avg 29903\n",
      "00:00:00    15 MB(-4%)  Iter   1  100.00%  K-mer dist pass 1\n",
      "00:00:00    15 MB(-4%)  Iter   1  100.00%  K-mer dist pass 2\n",
      "00:00:18  999 MB(-230%)  Iter   1  100.00%  Align node      \n",
      "00:00:18  999 MB(-230%)  Iter   1  100.00%  Root alignment\n",
      "\u001b[32m\u001b[22mRunning: \u001b[39m\u001b[22martic_get_stats --scheme /artic-ncov2019/primer_schemes/nCoV-2019/V3/nCoV-2019.scheme.bed --align-report sample.alignreport.txt --vcf-report sample.vcfreport.txt sample\n",
      "DeprecationWarning: 'source deactivate' is deprecated. Use 'conda deactivate'.\n",
      "00:00:00\tSnpEff version SnpEff 5.0c (build 2020-11-25 14:23), by Pablo Cingolani\n",
      "00:00:00\tCommand: 'ann'\n",
      "00:00:00\tReading configuration file 'snpEff.config'. Genome: 'NC_045512.2'\n",
      "00:00:00\tReading config file: /snpEff/snpEff.config\n",
      "00:00:00\tdone\n",
      "00:00:00\tReading database for genome version 'NC_045512.2' from file '/snpEff/./data/NC_045512.2/snpEffectPredictor.bin' (this might take a while)\n",
      "00:00:00\tDatabase not installed\n",
      "\tAttempting to download and install database 'NC_045512.2'\n",
      "00:00:00\tReading configuration file 'snpEff.config'. Genome: 'NC_045512.2'\n",
      "00:00:00\tReading config file: /snpEff/snpEff.config\n",
      "00:00:01\tdone\n",
      "00:00:01\tDownloading database for 'NC_045512.2'\n",
      "00:00:01\tConnecting to https://snpeff.blob.core.windows.net/databases/v5_0/snpEff_v5_0_NC_045512.2.zip\n",
      "00:00:02\tLocal file name: '/tmp/snpEff_v5_0_NC_045512.2.zip'\n",
      "\n",
      "00:00:02\tDonwload finished. Total 20946 bytes.\n",
      "00:00:02\tExtracting file 'data/NC_045512.2/snpEffectPredictor.bin'\n",
      "00:00:02\tCreating local directory: '/snpEff/./data/NC_045512.2'\n",
      "00:00:02\tUnzip: OK\n",
      "00:00:02\tDeleted local file '/tmp/snpEff_v5_0_NC_045512.2.zip'\n",
      "00:00:02\tDone\n",
      "00:00:02\tDatabase installed.\n",
      "00:00:02\tdone\n",
      "00:00:02\tLoading Motifs and PWMs\n",
      "00:00:02\tBuilding interval forest\n",
      "00:00:02\tdone.\n",
      "00:00:02\tGenome stats :\n",
      "#-----------------------------------------------\n",
      "# Genome name                : 'COVID19 Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1'\n",
      "# Genome version             : 'NC_045512.2'\n",
      "# Genome ID                  : 'NC_045512.2[0]'\n",
      "# Has protein coding info    : true\n",
      "# Has Tr. Support Level info : true\n",
      "# Genes                      : 11\n",
      "# Protein coding genes       : 11\n",
      "#-----------------------------------------------\n",
      "# Transcripts                : 38\n",
      "# Avg. transcripts per gene  : 3.45\n",
      "# TSL transcripts            : 0\n",
      "#-----------------------------------------------\n",
      "# Checked transcripts        : \n",
      "#               AA sequences :     12 ( 31.58% )\n",
      "#              DNA sequences :      0 ( 0.00% )\n",
      "#-----------------------------------------------\n",
      "# Protein coding transcripts : 38\n",
      "#              Length errors :      0 ( 0.00% )\n",
      "#  STOP codons in CDS errors :      0 ( 0.00% )\n",
      "#         START codon errors :     24 ( 63.16% )\n",
      "#        STOP codon warnings :     26 ( 68.42% )\n",
      "#              UTR sequences :      0 ( 0.00% )\n",
      "#               Total Errors :     24 ( 63.16% )\n",
      "# WARNING                    : No protein coding transcript has UTR\n",
      "#-----------------------------------------------\n",
      "# Cds                        : 40\n",
      "# Exons                      : 40\n",
      "# Exons with sequence        : 40\n",
      "# Exons without sequence     : 0\n",
      "# Avg. exons per transcript  : 1.05\n",
      "#-----------------------------------------------\n",
      "# Number of chromosomes      : 1\n",
      "# Chromosomes                : Format 'chromo_name size codon_table'\n",
      "#\t\t'NC_045512.2'\t29903\tStandard\n",
      "#-----------------------------------------------\n",
      "\n",
      "00:00:02\tPredicting variants\n",
      "\n",
      "ERRORS: Some errors were detected\n",
      "Error type\tNumber of errors\n",
      "ERROR_CHROMOSOME_NOT_FOUND\t2\n",
      "\n",
      "\n",
      "00:00:02\tCreating summary file: //sample.pass.snpeff.vcf.stats\n",
      "00:00:03\tCreating summary file: snpEff_summary.html\n",
      "00:00:03\tCreating genes file: //sample.pass.snpeff.vcf.stats.genes.txt\n",
      "00:00:03\tdone.\n",
      "00:00:03\tLogging\n",
      "00:00:04\tChecking for updates...\n",
      "00:00:05\tDone.\n",
      "pangolin already latest release (v2.3.2)\n",
      "pangoLEARN already latest release (2021-02-21)\n",
      "\u001b[32mFound the snakefile\u001b[0m\n",
      "\u001b[32mThe query file is:\u001b[0m/sample.consensus.fasta\n",
      "\u001b[36m\n",
      "--threads flag used, but threading not currently supported. Continuing with one thread.\u001b[0m\n",
      "Looking in /opt/conda/envs/pangolin/lib/python3.7/site-packages/pangoLEARN/data for data files...\n",
      "\u001b[32m\n",
      "Data files found\u001b[0m\n",
      "Trained model:\t/opt/conda/envs/pangolin/lib/python3.7/site-packages/pangoLEARN/data/decisionTree_v1.joblib\n",
      "Header file:\t/opt/conda/envs/pangolin/lib/python3.7/site-packages/pangoLEARN/data/decisionTreeHeaders_v1.joblib\n",
      "Lineages csv:\t/opt/conda/envs/pangolin/lib/python3.7/site-packages/pangoLEARN/data/lineages.metadata.csv\n",
      "\u001b[33mJob counts:\n",
      "\tcount\tjobs\n",
      "\t1\tadd_failed_seqs\n",
      "\t1\talign_to_reference\n",
      "\t1\tall\n",
      "\t1\tminimap2_check_distance\n",
      "\t1\toverwrite\n",
      "\t1\tpangolearn\n",
      "\t1\tparse_paf\n",
      "\t1\ttype_variants_b117\n",
      "\t1\ttype_variants_b1351\n",
      "\t1\ttype_variants_p1\n",
      "\t1\ttype_variants_p2\n",
      "\t11\u001b[0m\n",
      "\u001b[33mJob counts:\n",
      "\tcount\tjobs\n",
      "\t1\tparse_paf\n",
      "\t1\u001b[0m\n",
      "[M::mm_idx_gen::0.002*1.44] collected minimizers\n",
      "[M::mm_idx_gen::0.003*1.30] sorted minimizers\n",
      "[M::main::0.003*1.29] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.004*1.11] mid_occ = 100\n",
      "[M::mm_idx_stat] kmer size: 19; skip: 19; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.004*1.10] distinct minimizers: 2952 (100.00% are singletons); average occurrences: 1.000; average spacing: 10.130\n",
      "[M::worker_pipeline::0.015*1.03] mapped 1 sequences\n",
      "[M::main] Version: 2.17-r941\n",
      "[M::main] CMD: minimap2 -a -x asm5 -t 1 /opt/conda/envs/pangolin/lib/python3.7/site-packages/pangolin-2.3.2-py3.7.egg/pangolin/data/reference.fasta /tmp/tmpg7wl_xoe/mappable.fasta\n",
      "[M::main] Real time: 0.016 sec; CPU: 0.016 sec; Peak RSS: 0.005 GB\n",
      "warning: using --pad without --trim has no effect\n",
      "loading model 03/10/2021, 02:28:27\n",
      "processing block of 1 sequences 03/10/2021, 02:28:27\n",
      "complete 03/10/2021, 02:28:28\n",
      "\u001b[33mJob counts:\n",
      "\tcount\tjobs\n",
      "\t1\tadd_failed_seqs\n",
      "\t1\u001b[0m\n",
      "\u001b[33mJob counts:\n",
      "\tcount\tjobs\n",
      "\t1\toverwrite\n",
      "\t1\u001b[0m\n",
      "\u001b[32mOutput file written to: \u001b[0m/pangolin.csv\n",
      "DeprecationWarning: 'source deactivate' is deprecated. Use 'conda deactivate'.\n"
     ]
    }
   ],
   "source": [
    "#if os.getenv(\"ORBITER_SEQUENCING_PLATFORM\") == \"Oxford Nanopore\":\n",
    "\n",
    "mindepth = 50\n",
    "!bash covid19_call_variants.artic.sh {sample_filename} #> variants.log 2>&1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load reference genome\n",
    "#reference = list(SeqIO.parse(REFERENCE_PATH, \"fasta\"))\n",
    "#reference_length = len(reference[0])\n",
    "\n",
    "reference_length = 29903\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "!samtools depth $BAM_PATH > snps.depth 2> /dev/null"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#n_reads = sample.primary_classification.results()[\"n_reads\"]\n",
    "n_reads = 55000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "samtools_view_output = !samtools view -F 2308 $BAM_PATH | wc -l\n",
    "n_mapped_reads = int(samtools_view_output[0])\n",
    "percent_mapped_reads = n_mapped_reads/n_reads*100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "depth_table = []\n",
    "\n",
    "with open(\"snps.depth\") as handle:\n",
    "    for line in handle:\n",
    "        row = line.strip().split(\"\\t\")\n",
    "        depth_table.append(\n",
    "            {\"reference\": row[0], \"position\": int(row[1]), \"depth\": int(row[2])}\n",
    "        )\n",
    "depth_table = pd.DataFrame(depth_table, columns=[\"reference\", \"position\", \"depth\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate genome coverage (what percent of bases are coveraged at X coverage)\n",
    "# Use a fixed reference length that we use for `samtools depth` above\n",
    "\n",
    "covered_sites = set()\n",
    "covered_sites_mindepth = set()\n",
    "\n",
    "for _, row in depth_table.iterrows():\n",
    "    row = row.to_dict()\n",
    "    if row[\"depth\"] >= 1:\n",
    "        covered_sites.add(row[\"position\"])\n",
    "    if row[\"depth\"] >= mindepth:\n",
    "        covered_sites_mindepth.add(row[\"position\"])        \n",
    "\n",
    "cov = len(covered_sites) / reference_length\n",
    "cov_mindepth = len(covered_sites_mindepth) / reference_length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get mean over windows because altair can't handle > 5k points ...\n",
    "binned_depths = []\n",
    "window_width = reference_length // 4500\n",
    "\n",
    "for i in range(1, reference_length, window_width):\n",
    "    window = depth_table.loc[\n",
    "        (depth_table[\"position\"] > i) & (depth_table[\"position\"] < i + window_width)\n",
    "    ]\n",
    "\n",
    "    binned_depths.append(\n",
    "        {\"position\": i, \"depth\": window[\"depth\"].mean(),}\n",
    "    )\n",
    "\n",
    "binned_depths = pd.DataFrame(binned_depths)\n",
    "mean_depth = depth_table[\"depth\"].mean() if not depth_table.empty else 0\n",
    "median_depth = depth_table[\"depth\"].median() if not depth_table.empty else 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read from variants.tsv\n",
    "snp_table = pd.read_csv(VARIANTS_TSV_PATH, sep=\"\\t\")\n",
    "# Read Nextclade and Pangolin tables\n",
    "nextclade_table = pd.read_csv(NEXTCLADE_TSV_PATH, sep=\"\\t\")\n",
    "pangolin_table = pd.read_csv(PANGOLIN_CSV_PATH, sep=\",\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_snps = snp_table.shape[0]\n",
    "n_snps_mindepth = snp_table[snp_table['depth']>=mindepth].shape[0]\n",
    "nextclade_lineage = nextclade_table['clade'].iloc[0]\n",
    "nextclade_pm_count = nextclade_table['qc.privateMutations.total'].iloc[0]\n",
    "pangolin_lineage = pangolin_table['lineage'].iloc[0]\n",
    "pangolin_version = pangolin_table['pangoLEARN_version'].iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h2 class=\"title\" style=\"\">SARS-CoV-2 (COVID-19) Sequencing Overview</h2>"
      ],
      "text/plain": [
       "<onecodex.notebooks.report.title at 0x7fea77c92a60>"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "title(\"SARS-CoV-2 (COVID-19) Sequencing Overview\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "This report summarizes the detection of SARS-CoV-2 in sample \n",
       "<strong>sample.fastq</strong>. \n",
       "\n",
       "<p>This sample contained <strong>55,000</strong> reads, with\n",
       "<strong>244704.9%</strong> mapping to the \n",
       "<a href='https://www.ncbi.nlm.nih.gov/nuccore/MN908947.3/' target='_blank'>Wuhan-Hu-1 reference</a>.\n",
       "Reads span <strong>98%</strong> of the genome, with 95% of the genome covered at depths >50x, with a mean depth of <strong>4012x</strong>.</p>\n",
       "\n",
       "<p>A total of <strong>2</strong> variants were detected at depths over 50x.\n",
       "This genome is classified as Pangolin lineage <strong>B.1</strong> using PangoLEARN verison 2021-02-21 and Nextclade lineage <strong>19A</strong> with 1 private mutation.</p>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "text = f\"\"\"\n",
    "This report summarizes the detection of SARS-CoV-2 in sample \n",
    "<strong>{sample_filename}</strong>. \n",
    "\n",
    "<p>This sample contained <strong>{n_reads:,}</strong> reads, with\n",
    "<strong>{percent_mapped_reads:.1%}</strong> mapping to the \n",
    "<a href='https://www.ncbi.nlm.nih.gov/nuccore/MN908947.3/' target='_blank'>Wuhan-Hu-1 reference</a>.\n",
    "Reads span <strong>{cov:.0%}</strong> of the genome, with {cov_mindepth:.0%} of the genome covered at depths >{mindepth:}x, with a mean depth of <strong>{mean_depth:.0f}x</strong>.</p>\n",
    "\n",
    "<p>A total of <strong>{n_snps_mindepth}</strong> variant{'s were' if n_snps_mindepth != 1 else 'was'} detected at depths over {mindepth:}x.\n",
    "This genome is classified as Pangolin lineage <strong>{pangolin_lineage}</strong> using PangoLEARN verison {pangolin_version} and Nextclade lineage <strong>{nextclade_lineage}</strong> with {nextclade_pm_count} private mutation{'s' if nextclade_pm_count != 1 else ''}.</p>\"\"\"\n",
    "\n",
    "HTML(text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "internal/modules/cjs/loader.js:965\n",
      "  throw err;\n",
      "  ^\n",
      "\n",
      "Error: Cannot find module 'vega'\n",
      "Require stack:\n",
      "- /usr/lib/node_modules/vega-lite/build/vega-lite.js\n",
      "- /usr/lib/node_modules/vega-lite/bin/vl2vg\n",
      "    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:962:15)\n",
      "    at Function.Module._load (internal/modules/cjs/loader.js:838:27)\n",
      "    at Module.require (internal/modules/cjs/loader.js:1022:19)\n",
      "    at require (internal/modules/cjs/helpers.js:72:18)\n",
      "    at /usr/lib/node_modules/vega-lite/build/vega-lite.js:2:105\n",
      "    at Object.<anonymous> (/usr/lib/node_modules/vega-lite/build/vega-lite.js:5:2)\n",
      "    at Module._compile (internal/modules/cjs/loader.js:1118:30)\n",
      "    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1138:10)\n",
      "    at Module.load (internal/modules/cjs/loader.js:982:32)\n",
      "    at Function.Module._load (internal/modules/cjs/loader.js:875:14) {\n",
      "  code: 'MODULE_NOT_FOUND',\n",
      "  requireStack: [\n",
      "    '/usr/lib/node_modules/vega-lite/build/vega-lite.js',\n",
      "    '/usr/lib/node_modules/vega-lite/bin/vl2vg'\n",
      "  ]\n",
      "}\n"
     ]
    },
    {
     "ename": "CalledProcessError",
     "evalue": "Command '['/usr/bin/vl2vg']' returned non-zero exit status 1.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/IPython/core/formatters.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, obj, include, exclude)\u001b[0m\n\u001b[1;32m    968\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    969\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mmethod\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 970\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minclude\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0minclude\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexclude\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mexclude\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    971\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    972\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/altair/vegalite/v4/api.py\u001b[0m in \u001b[0;36m_repr_mimebundle_\u001b[0;34m(self, include, exclude)\u001b[0m\n\u001b[1;32m   1459\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1460\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1461\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mrenderers\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdct\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1462\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1463\u001b[0m     def display(self, renderer=Undefined, theme=Undefined, actions=Undefined,\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/toolz/functoolz.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    301\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__call__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    302\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 303\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_partial\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    304\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mTypeError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mexc\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    305\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_should_curry\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwargs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/altair_saver/_core.py\u001b[0m in \u001b[0;36mrender\u001b[0;34m(chart, fmts, mode, embed_options, method, **kwargs)\u001b[0m\n\u001b[1;32m    255\u001b[0m         \u001b[0mSaver\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_select_saver\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmode\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfmt\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfmt\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    256\u001b[0m         \u001b[0msaver\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSaver\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspec\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmode\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0membed_options\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0membed_options\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 257\u001b[0;31m         \u001b[0mmimebundle\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msaver\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmimebundle\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfmt\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    258\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    259\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mmimebundle\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/altair_saver/savers/_saver.py\u001b[0m in \u001b[0;36mmimebundle\u001b[0;34m(self, fmts)\u001b[0m\n\u001b[1;32m     88\u001b[0m                 \u001b[0mvegalite_version\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_package_versions\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"vega-lite\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     89\u001b[0m             )\n\u001b[0;32m---> 90\u001b[0;31m             \u001b[0mbundle\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mmimetype\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_serialize\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfmt\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"mimebundle\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     91\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mbundle\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     92\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/altair_saver/savers/_node.py\u001b[0m in \u001b[0;36m_serialize\u001b[0;34m(self, fmt, content_type)\u001b[0m\n\u001b[1;32m    112\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    113\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_mode\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"vega-lite\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 114\u001b[0;31m             \u001b[0mspec\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_vl2vg\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspec\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    115\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    116\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mfmt\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"vega\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/altair_saver/savers/_node.py\u001b[0m in \u001b[0;36m_vl2vg\u001b[0;34m(self, spec)\u001b[0m\n\u001b[1;32m     63\u001b[0m         \u001b[0mvl2vg\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mexec_path\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"vl2vg\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     64\u001b[0m         \u001b[0mvl_json\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mjson\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdumps\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspec\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mencode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 65\u001b[0;31m         vg_json = check_output_with_stderr(\n\u001b[0m\u001b[1;32m     66\u001b[0m             \u001b[0;34m[\u001b[0m\u001b[0mvl2vg\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minput\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mvl_json\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstderr_filter\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_stderr_filter\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     67\u001b[0m         )\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/site-packages/altair_saver/_utils.py\u001b[0m in \u001b[0;36mcheck_output_with_stderr\u001b[0;34m(cmd, shell, input, stderr_filter)\u001b[0m\n\u001b[1;32m    196\u001b[0m     \"\"\"\n\u001b[1;32m    197\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 198\u001b[0;31m         ps = subprocess.run(\n\u001b[0m\u001b[1;32m    199\u001b[0m             \u001b[0mcmd\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    200\u001b[0m             \u001b[0mshell\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mshell\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.8/subprocess.py\u001b[0m in \u001b[0;36mrun\u001b[0;34m(input, capture_output, timeout, check, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    510\u001b[0m         \u001b[0mretcode\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpoll\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    511\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mcheck\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mretcode\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 512\u001b[0;31m             raise CalledProcessError(retcode, process.args,\n\u001b[0m\u001b[1;32m    513\u001b[0m                                      output=stdout, stderr=stderr)\n\u001b[1;32m    514\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mCompletedProcess\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mretcode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstdout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstderr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mCalledProcessError\u001b[0m: Command '['/usr/bin/vl2vg']' returned non-zero exit status 1."
     ]
    },
    {
     "data": {
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Coverage plot\n",
    "plot = (\n",
    "    alt.Chart(binned_depths)\n",
    "    .mark_area()\n",
    "    .transform_window(rolling_mean=\"mean(depth)\", frame=[-50, 50])\n",
    "    .encode(\n",
    "        x=alt.X(\n",
    "            \"position\",\n",
    "            title=\"Genomic Coordinate\",\n",
    "            scale=alt.Scale(domain=[0, reference_length]),\n",
    "        ),\n",
    "        y=alt.Y(\"rolling_mean:Q\", scale=alt.Scale(type=\"linear\"), title=\"Depth\"),\n",
    "    )\n",
    "    .properties(\n",
    "        title=f\"SARS-CoV-2 ({REFERENCE_NAME})\",\n",
    "        width=550,\n",
    "        height=150,\n",
    "    )\n",
    ")\n",
    "plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Position</th>\n",
       "      <th>Variant</th>\n",
       "      <th>Depth</th>\n",
       "      <th>Frequency</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Protein mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>14408</td>\n",
       "      <td>C → T</td>\n",
       "      <td>357</td>\n",
       "      <td>96.08%</td>\n",
       "      <td>nsp12 in orf1ab</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>23403</td>\n",
       "      <td>A → G</td>\n",
       "      <td>351</td>\n",
       "      <td>97.72%</td>\n",
       "      <td>spike</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Can use the previously defined mindepth instead of DEPTH_FILTER\n",
    "\n",
    "if snp_table.empty or snp_table.loc[snp_table[\"depth\"] >= mindepth, :].shape[0] == 0:\n",
    "    table = pd.DataFrame(columns=[\"POS\", \"Variant\", \"Gene\"])\n",
    "    # should display mindepth instead of 10x\n",
    "    table = HTML(\n",
    "        \"<div style='text-align: center; color: #555; width: 100%;'><em>No SNPs found >= 50x depth.</em><br /><br /></div>\"\n",
    "    )\n",
    "else:\n",
    "    snp_table[\"Position\"] = snp_table[\"position\"]\n",
    "    snp_table[\"Variant\"] = [f\"{r['ref allele']} → {r['alt allele']}\" for _, r in snp_table.iterrows()]\n",
    "    snp_table[\"Depth\"] = snp_table[\"depth\"]\n",
    "    snp_table[\"Frequency\"] = snp_table[\"alt frequency\"] # Already expressed as a percentage to 2 decimal places\n",
    "    snp_table[\"Gene\"] = snp_table[\"orf\"]\n",
    "    snp_table[\"Protein mutation\"] = snp_table[\"protein sequence variant\"]\n",
    "\n",
    "    table = snp_table.loc[\n",
    "        snp_table[\"Depth\"] >= mindepth, [\"Position\", \"Variant\", \"Depth\", \"Frequency\", \"Gene\", \"Protein mutation\"]\n",
    "    ]  # , \"Gene\", \"Protein change\"]]\n",
    "    table = HTML(table.to_html(index=False))\n",
    "table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"text-align: center; padding-top: 10px; font-size: 0.7em; color: #777;\"><em>SARS-CoV-2 variants.</em></div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "legend_text = \"SARS-CoV-2 variants.\"\n",
    "n_extra_variants = (\n",
    "    snp_table[snp_table[\"Depth\"] < mindepth].shape[0] if not snp_table.empty else 0\n",
    ")\n",
    "if n_extra_variants:\n",
    "    legend_text += f\" An additional {n_extra_variants} variant{'s' if n_extra_variants > 1 else ''} <{mindepth}× depth {'are' if n_extra_variants > 1 else 'is'} not shown.\"\n",
    "if os.environ.get(\"ONE_CODEX_REPORT_UUID\"):\n",
    "    legend_text += f\"\"\" \n",
    "         A variants TSV and consensus FASTA is available <a target=\"_blank\" href=\\\"{'https://app.onecodex.com/report/' + os.environ['ONE_CODEX_REPORT_UUID'] + '/files'}\\\">here</a>.\n",
    "        \"\"\"\n",
    "HTML(\n",
    "    '<div style=\"text-align: center; padding-top: 10px; font-size: 0.7em; color: #777;\"><em>'\n",
    "    + legend_text\n",
    "    + \"</em></div>\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Additional Resources\n",
    "\n",
    "- Additional bioinformatics pipeline details are [available on GitHub](https://github.com/onecodex/sars-cov-2)\n",
    "- [Nextstrain](https://nextstrain.org/ncov) maintains an up-to-date analysis of SARS-CoV-2 (HCoV-19).\n",
    "- The [Global Initiative on Sharing All Influenza Data (GISAID)](https://www.gisaid.org/) hosts viral genomes from ongoing outbreaks. Please [contact us](mailto:hello@onecodex.com) for help submitting your data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<style type='text/css'>\n",
       "@page {\n",
       "    @bottom-center {\n",
       "        content: \" NOT FOR DIAGNOSTIC USE\" !important;\n",
       "    }\n",
       "}\n",
       "</style>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add One Codex report ID to footer for reproducibility/data provenance (not yet in v0.7.2)\n",
    "HTML(\n",
    "    f\"\"\"\n",
    "<style type='text/css'>\n",
    "@page {{\n",
    "    @bottom-center {{\n",
    "        content: \"{os.environ['ONE_CODEX_REPORT_UUID'] + ' -' if os.environ.get('ONE_CODEX_REPORT_UUID') else ''} NOT FOR DIAGNOSTIC USE\" !important;\n",
    "    }}\n",
    "}}\n",
    "</style>\n",
    "\"\"\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'sample' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-69-08ae85879f83>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     12\u001b[0m }\n\u001b[1;32m     13\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 14\u001b[0;31m \u001b[0;32mwith\u001b[0m \u001b[0mgzip\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"{sample.filename}.report.json.gz\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"w\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     15\u001b[0m     \u001b[0mf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mjson\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdumps\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mencode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'sample' is not defined"
     ]
    }
   ],
   "source": [
    "# Save a JSON too, including filtered variants <50x\n",
    "results = {\n",
    "    \"n_reads\": n_reads,\n",
    "    \"n_mapped_reads\": n_mapped_reads,\n",
    "    \"report_id\": os.environ.get(\"ONE_CODEX_REPORT_UUID\"), \n",
    "    \"sample_id\": os.environ.get(\"ONE_CODEX_SAMPLE_UUID\"),\n",
    "    \"variants\": [r.to_dict() for _, r in snp_table.iterrows()],\n",
    "    \"coverage\": cov,\n",
    "    \"coverage_over_50x\": cov_mindepth,\n",
    "    \"mean_depth\": mean_depth,\n",
    "    \"median_depth\": median_depth,\n",
    "}\n",
    "\n",
    "with gzip.open(f\"{sample.filename}.report.json.gz\", \"w\") as f:\n",
    "    f.write(json.dumps(results).encode())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clean up files\n",
    "!rm -f {sample.filename} snps.depth variants.log covid19.bam.bai"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
